server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java

```diff
@@ -283,11 +322,149 @@ private void listStartedShards(
                 ActionListener.wrap(listener::onResponse, listener::onFailure)
             );
         }
+
+        Lister<BaseNodesResponse<NodeGatewayStartedShards>, NodeGatewayStartedShards> batchLister = (
+            shardId,
+            customDataPath,
+            nodes,
+            listener) -> {
+            // group shards by node
+            for (DiscoveryNode node : nodes) {
+                List<ShardRequestInfo<NodeGatewayStartedShards>> nodeLevelRequests = queuedRequests.computeIfAbsent(
+                    node,
+                    n -> Collections.synchronizedList(new ArrayList<>())
+                );
+                nodeLevelRequests.add(new ShardRequestInfo<>(shardId, customDataPath, listener));
+
+                synchronized (listenerMutex) {
+                    // queue child listeners
+                    List<ActionListener<BaseNodesResponse<NodeGatewayStartedShards>>> shardListeners = queuedListeners.computeIfAbsent(
```

Comment:
Hi @howardhuanghua, the size of `queuedListeners` is O(shard) * O(dataNode),   and  it may takes up a lot of memory space When the cluster has more shards and data-nodes. Besides，many data fetch cache in asyncFetchData cannot be removed or cleared due to it need all the datanodes successful response.

Reply:
Hi @GrandFisher Thanks for the review.
1. Before receiving node level shards fetch results, the queued listeners are empty `ArrayList` elements. At the beginning of full cluster restart, node resource should be enough.
2. The queued node shard listeners would be cleared immediately once receiving the node level response.
3. Alternately, we could separate the shard requests in several of batches instead of flush all the shards in a big batch.

Reply:
After double checking, `queuedListerners` do have the risk to take up a lot of memory space.
I replaced `DiscoveryNode` with node id in single action listener. We will verify the memory consumption again. Thanks.